doc/sg_zone.8
.TH SG_ZONE "8" "2026-07-01" "sg3_utils\-1.49" SG3_UTILS
.SH NAME
sg_zone \- send a SCSI ZONE modifying command
.SH SYNOPSIS
.B sg_zone
[\fI\-\-all\fR] [\fI\-\-close\fR] [\fI\-\-count=ZC\fR] [\fI\-\-element=EID\fR]
[\fI\-\-finish\fR] [\fI\-\-help\fR] [\fI\-\-open\fR] [\fI\-\-remove\fR]
[\fI\-\-sequentialize\fR] [\fI\-\-timeout=SE\fR] [\fI\-\-verbose\fR]
[\fI\-\-version\fR] [\fI\-\-zone=ID\fR] \fIDEVICE\fR
.SH DESCRIPTION
.\" Add any additional description here
Sends a SCSI OPEN ZONE, CLOSE ZONE, FINISH ZONE, REMOVE ELEMENT AND MODIFY
ZONES or SEQUENTIALIZE ZONE command to the \fIDEVICE\fR. All but the last
two are found in the ZBC standard (INCITS 536\-2016). The REMOVE ELEMENT AND
MODIFY ZONES command was added in zbc2r07 while the SEQUENTIALIZE ZONE command
was added in zbc2r01b.
.PP
One and only one of the \fI\-\-open\fR, \fI\-\-close\fR, \fI\-\-finish\fR,
\fI\-\-remove\fR and \fI\-\-sequentialize\fR options can be chosen.
.PP
The REPORT ZONES, REPORT REALMS and REPORT ZONE DOMAINS commands may be
accessed via the sg_rep_zones utility. The ZONE ACTIVATE and ZONE QUERY
commands may be accessed via the sg_z_act_query utility. The RESET WRITE
POINTER command may be accessed via the sg_reset_wp utility.
.SH OPTIONS
Arguments to long options are mandatory for short options as well.
.TP
\fB\-a\fR, \fB\-\-all\fR
sets the ALL field in the cdb.
.TP
\fB\-c\fR, \fB\-\-close\fR
causes the CLOSE ZONE command to be sent to the \fIDEVICE\fR.
.TP
\fB\-C\fR, \fB\-\-count\fR=\fIZC\fR
ZC is placed in the Zone Count field in the cdb of all four commands
supported by this utility. ZC should be a value from 0 to 65535 (0xffff)
inclusive.
.TP
\fB\-e\fR, \fB\-\-element\fR=\fIEID\fR
where \fIEID\fR is an element identifier which is a 32 bit unsigned integer
starting at one. This field is used by the REMOVE ELEMENT AND MODIFY ZONES
command and its default value is zero (which is invalid). So the user needs
to supply a valid element identifier when \fI\-\-remove\fR is used.
.TP
\fB\-f\fR, \fB\-\-finish\fR
causes the FINISH ZONE command to be sent to the \fIDEVICE\fR.
.TP
\fB\-h\fR, \fB\-\-help\fR
output the usage message then exit.
.TP
\fB\-o\fR, \fB\-\-open\fR
causes the OPEN ZONE command to be sent to the \fIDEVICE\fR.
.TP
\fB\-r\fR, \fB\-\-remove\fR
causes the REMOVE ELEMENT AND MODIFY ZONES command to be sent to the
\fIDEVICE\fR. In practice, \fI\-\-element=EID\fR needs to be also given.
.TP
\fB\-S\fR, \fB\-\-sequentialize\fR
causes the SEQUENTIALIZE ZONE command to be sent to the \fIDEVICE\fR.
.TP
\fB\-t\fR, \fB\-\-timeout\fR=\fISE\fR
where \fISE\fR is the command timeout in seconds. The default is 60 seconds
and if 0 is given, it is mapped to 60. An alternate form is \fI\-\-tmo=SE\fR.
.TP
\fB\-v\fR, \fB\-\-verbose\fR
increase the level of verbosity, (i.e. debug output).
.TP
\fB\-V\fR, \fB\-\-version\fR
print the version string and then exit.
.TP
\fB\-z\fR, \fB\-\-zone\fR=\fIID\fR
where \fIID\fR is placed in the cdb's ZONE ID field. A zone id is a zone
start logical block address (LBA). The default value is 0. \fIID\fR is
assumed to be in decimal unless prefixed with '0x' or has a trailing 'h'
which indicate hexadecimal.
.SH NOTES
After a REMOVE ELEMENT AND MODIFY ZONES command has completed, the element
in question is said to be depopulated and any affected zones are placed in
the 'offline' zone condition.
.PP
SBC\-4 has a similar command to REMOVE ELEMENT AND MODIFY ZONES called REMOVE
ELEMENT AND TRUNCATE. The difference is that the latter "changes the
association between LBAs and physical blocks" and the former does not change
that association. In both cases, depopulated elements that have
the 'Restoration Allowed' (RALWD) bit set (see sg_get_elem_status) may be
restored with the RESTORE ELEMENTS AND REBUILD command, see the
sg_rem_rest_elem utility. The sg_rem_rest_elem utility can also issue the
REMOVE ELEMENT AND MODIFY ZONES command with its \fI\-\-modify\fR option.
.SH EXIT STATUS
The exit status of sg_zone is 0 when it is successful. Otherwise see
the sg3_utils(8) man page.
.SH AUTHORS
Written by Douglas Gilbert.
.SH "REPORTING BUGS"
Report bugs to <dgilbert at interlog dot com>.
.SH COPYRIGHT
Copyright \(co 2014\-2026 Douglas Gilbert
.br
This software is distributed under a BSD\-2\-Clause license. There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
.SH "SEE ALSO"
.B sg_rem_rest_elem,sg_rep_zones,sg_reset_wp,sg_z_act_query(sg3_utils)


ChangeLog
Each utility has its own version number, date of last change and
some description at the top of its ".c" file. All utilities in the main
directory have their own "man" pages. There is also a sg3_utils man page.
References to github are to https://github.com/doug-gilbert/sg3_utils

Changelog for pre-release sg3_utils-1.49 [20260701] [svn: r1131]
  Utility specific changes:
  - rescan-scsi-bus.sh: fix multipath resize without update 
    https://github.com/doug-gilbert/sg3_utils/pull/43.diff
    and 44.diff
    - fix for github.com/doug-gilbert/sg3_utils/issues/46
    - github PR numbers: 68, 69, 71 and 74 by Paul Evans
    - github pull req: #72 fix udevadm invocation M. Wilck
    - github issue: #76 fix shellcheck warnings Xose Vazquez Perez
  - scsi_logging_level: add MIT license Xose Vazquez Perez
  - scsi-rescan.8 manpage added, links to rescan-scsi-bus.8
  - sg_copy_results: part of github PR #74
  - sg_dd: add grpnum=GN operand and --nocopy option
  - sg_get_elem_status: fractional elements, sbc6r02
    - correct several error messages
  - sg_get_lba_status: additional LBA accessibility values
  - sginfo: part of github PR #74
  - sg_inq: update version descriptors to T10 table 20230814
  - sg_inq+sg_vpd: show depopulation time in hours and
    minutes as well as its raw form: seconds
    - add RSCS field to Block limits ext. vpage [sbc5r5]
    - translation of NVMe FR field uses sg_last_n_non_blank()
    - add NVMe Information VPD page
    - clean up ATA Information VPD page
    - rework designation descriptor handling
    - sg_inq --export conformance for SCSI name string
      by Paul Evans github PR #83
  - sg_lib: add sg_last_n_non_blank() for SNT FR field
    - sg_lib.[hc] add sg_strscpy() to replace strncpy()
    - sg_json.c fixes from lsscsi project pull #15
    - sg_json.[hc] add sgj_haj_vs_hex_bytes_nex() for
      outputing IDs as ASCII hex strings
    - sg_json.[hc] add sgj_haj_vs_len()
    - sg_json.[hc] add sgj_haj_vs_nex()
    - sg_json_sg_lib.[hc] sgj_js_designation_descriptor()
      changed to sgj_haj_designation_descriptor()
  - sg_logs: correct tape Device statisitics lpage [0x14]
    serial number parameters [0x40 and 0x41] per
    https://github.com/doug-gilbert/sg3_utils/issues/45
  - sg_map: handle LUNs > 255
  - sg_opcodes: fix rwcdl/cdlp fields handling D. Le Moal
  - sg_rem_rest_elem: add support for REMOVE ELEMENT AND
  - MODIFY ZONES command, Damien Le Moal
  - sg_requests: improve buffer truncation
  - sg_ses: restore some help text, github PR #79 Paul Evans
  - sg_stream_ctl: add --hex, --inhex=, --json= and --raw
    options; additions in sbc5r05 (perm and relative_lifetime
    fields).
  - sg_turs: fix missing CDB on several codepaths
  - sg_vpd: inform about SPC-6 (sg_inq knew already)
    https://github.com/doug-gilbert/sg3_utils/issues/40
    - add --descriptors option, for version descriptors
    - part of github PR #74
    - some sg_vpd fixes shown under sg_inq+sg_vpd entry above
  - sg_wr_mode: fix --contents and --cfile= handling
    - add --mfile=MF option for decoding mask in the MF file
    - fix introduced Makefile.am bug, plus only for Linux
  - sg_xcopy: part of github PR #74
  - utils/hxascdmp: sprint -> snprintf [github PR #78]

  General changes:
  - fix 'make distcheck' (again)
  - add cmake build system (alongside autotools) with cpack
    for deb and rpm packages; ctest not wired yet
    - github PR #80 by Alan Somers which is associated
      with cpack FreeBSD settings
    - github PR #82 by Paul Evans to get cmake output to be
      equivalent (in parity) with what Autotools produces
    - wire option(DISABLE_LIBSGUTILS (def: OFF)) so when
      ON sgutils2 library is static but other libraries
      are shared objects (e.g. libc.so)
    - github PR #84 by Paul Evans to fix cpack generated
      Debian dev library path, needed by sdparm and ddpt
  - fixes so musl-gcc builds clean
  - start complying with T10 23-047r1 "Broadcom SNT Reference"
    document assuming T10 SNT drafts are coming
  - rearrange SNT and NVMe headers and associated source
  - new error code: SG_LIB_PROGRESS_NOT_READY for long
    duration 'in progress' additional sense codes
  - lib/sg_pt_solaris.c : fix compilation issue per
    https://github.com/doug-gilbert/sg3_utils/pull/41
  - lib/sgj_haj_*() functions: tweak hex_haj argument meaning
  - scripts/00-scsi-sg3_config.rules : apply
    https://github.com/doug-gilbert/sg3_utils/pull/42
  - rename build_debian.sh script to debian_build.sh
  - rename testing directory to ttesting to avoid clash with
    cmake's CTEST that creates a Testing directory
  - ttesting/sg_chk_inq_vd.c: test internal table against T10
    version descriptor file
  - check/clean against gcc-16 -fanalyzer; clang-22 --analyze
  - debian build: modernize with dh sequencer [github PR# 73]
  - AIX partial support: needs getopt.h header which
    surprisingly is non-standard. Keep hidden version in
    lib directory and copy to include dir as needed
     - still no real pass-through, using sg_dummy instead
  - getopt_long directory contents -> lib directory
  - JSON: make output more consistent so most command
    responses have a *_paramter_data or similar sub-object
  - apply https://github.com/doug-gilbert/sg3_utils/pull/39
    and its revision [20230807] mainly for Android

Changelog for released sg3_utils-1.48 [20230801] [svn: r1042]
  - decoding utilities: add --json[=JO] and --js-file=JFN
    options, short form changed to -j[=JO]
  - sg_sat_datetime: new utility for setting or accessing
    the data and time on a ATA device using a SAT layer.
  - sg_z_act_query: new utility for sending either a
    Zone activate or Zone query command
  - sg_rep_density: new utility for decoding the response of
    Report density support command [ssc (tape)]
  - sg_rem_rest_elem: new utility for removing or restoring
    elements
  - sg_write_attr: new utility, Write Attribute is a SPC
    command that is used mainly by tape drives (ssc)

src/sg_zone.c
/*
 * Copyright (c) 2014-2026 Douglas Gilbert.
 * All rights reserved.
 *
 * SPDX-License-Identifier: BSD-2-Clause
 */

#include <unistd.h>
#include <fcntl.h>
#include <stdio.h>
#include <stdlib.h>
#include <stdarg.h>
#include <stdbool.h>
#include <string.h>
#include <ctype.h>
#include <getopt.h>
#define __STDC_FORMAT_MACROS 1
#include <inttypes.h>

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#include "sg_lib.h"
#include "sg_lib_data.h"
#include "sg_pt.h"
#include "sg_cmds_basic.h"
#include "sg_unaligned.h"
#include "sg_pr2serr.h"

/* A utility program originally written for the Linux OS SCSI subsystem.
 *
 * This program issues one of the following SCSI commands:
 *   - CLOSE ZONE
 *   - FINISH ZONE
 *   - OPEN ZONE
 *   - REMOVE ELEMENT AND MODIFY ZONES
 *   - SEQUENTIALIZE ZONE
 */

static const char * version_str = "1.22 20260701";

#define SG_ZONING_OUT_CMDLEN 16
#define CLOSE_ZONE_SA 0x1
#define FINISH_ZONE_SA 0x2
#define OPEN_ZONE_SA 0x3
#define SEQUENTIALIZE_ZONE_SA 0x10
#define REM_ELEM_MOD_ZONES_SA 0x1a      /* uses SERVICE ACTION IN(16) */

#define SENSE_BUFF_LEN 64       /* Arbitrary, could be larger */
#define DEF_PT_TIMEOUT  60      /* 60 seconds */


static const struct option long_options[] = {
    {"all", no_argument, 0, 'a'},
    {"close", no_argument, 0, 'c'},
    {"count", required_argument, 0, 'C'},
    {"element", required_argument, 0, 'e'},
    {"finish", no_argument, 0, 'f'},
    {"help", no_argument, 0, 'h'},
    {"open", no_argument, 0, 'o'},
    {"quick", no_argument, 0, 'q'},
    {"remove", no_argument, 0, 'r'},
    {"reset-all", no_argument, 0, 'R'},     /* same as --all */
    {"reset_all", no_argument, 0, 'R'},
    {"sequentialize", no_argument, 0, 'S'},
    {"timeout", required_argument, 0, 't'},
    {"tmo", required_argument, 0, 't'},
    {"verbose", no_argument, 0, 'v'},
    {"version", no_argument, 0, 'V'},
    {"zone", required_argument, 0, 'z'},
    {0, 0, 0, 0},
};

/* Indexed by service action of opcode 0x94 (Zone out) unless noted */
static const char * sa_name_arr[] = {
    "no SA=0",                  /* 0x0 */
    "Close zone",
    "Finish zone",
    "Open zone",
    "-", "-", "-", "-",
    "-",
    "-", "-", "-", "-",
    "-",
    "-",
    "-",
    "Sequentialize zone",       /* 0x10 */
    "-", "-", "-", "-",
    "-", "-", "-", "-",
    "-",
    "Remove element and modify zones", /* service action in(16), 0x1a */
};


static void
usage()
{
    pr2serr("Usage: "
            "sg_zone  [--all] [--close] [--count=ZC] [--element=EID] "
            "[--finish]\n"
            "                [--help] [--open] [--quick] [--remove] "
            "[--sequentialize]\n"
            "                [--timeout=SE] [--verbose] [--version] "
            "[--zone=ID]\n"
            "                DEVICE\n");
    pr2serr("  where:\n"
            "    --all|-a           sets the ALL flag in the cdb\n"
            "    --close|-c         issue CLOSE ZONE command\n"
            "    --count=ZC|-C ZC    set zone count field (def: 0)\n"
            "    --element=EID|-e EID    EID is the element identifier to "
            "remove;\n"
            "                            default is 0 which is an invalid "
            "EID\n"
            "    --finish|-f        issue FINISH ZONE command\n"
            "    --help|-h          print out usage message\n"
            "    --open|-o          issue OPEN ZONE command\n"
            "    --quick|-q         bypass 15 second warn and wait "
            "(for --remove)\n"
            "    --remove|-r        issue REMOVE ELEMENT AND MODIFY ZONES "
            "command\n"
            "    --sequentialize|-S    issue SEQUENTIALIZE ZONE command\n"
            "    --timeout=SE|-t SE    command timeout in seconds (def: "
            "60 secs)\n"
            "    --verbose|-v       increase verbosity\n"
            "    --version|-V       print version string and exit\n"
            "    --zone=ID|-z ID    ID is the starting LBA of the zone "
            "(def: 0)\n\n"
            "Performs a SCSI OPEN ZONE, CLOSE ZONE, FINISH ZONE, "
            "REMOVE ELEMENT AND\nMODIFY ZONES or SEQUENTIALIZE ZONE "
            "command. Either --close, --finish,\n--open, --remove or "
            "--sequentialize option needs to be given.\n");
}

/* Invokes the zone out command indicated by 'sa' (ZBC).  Return of 0
 * -> success, various SG_LIB_CAT_* positive values or -1 -> other errors */
static int
sg_ll_zone_out(int sg_fd, int sa, uint64_t zid, uint16_t zc, bool all,
               int tmo, bool noisy, int verbose)
{
    int ret, res, sense_cat;
    struct sg_pt_base * ptvp;
    uint8_t zo_cdb[SG_ZONING_OUT_CMDLEN] =
          {SG_ZONING_OUT, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,  0, 0, 0, 0};
    uint8_t sense_b[SENSE_BUFF_LEN] SG_C_CPP_ZERO_INIT;
    char b[64];

    zo_cdb[1] = 0x1f & sa;
    if (REM_ELEM_MOD_ZONES_SA == sa) {  /* zid carries element identifier */
        zo_cdb[0] = SG_SERVICE_ACTION_IN_16;    /* N.B. changing opcode */
        sg_put_unaligned_be32((uint32_t)zid, zo_cdb + 10); /* element id */
    } else {
        sg_put_unaligned_be64(zid, zo_cdb + 2);
        sg_put_unaligned_be16(zc, zo_cdb + 12);
        if (all)
            zo_cdb[14] = 0x1;
    }
    sg_get_opcode_sa_name(zo_cdb[0], sa, -1, sizeof(b), b);
    if (verbose) {
        char d[128];

        pr2serr("    %s cdb: %s\n", b,
                sg_get_command_str(zo_cdb, SG_ZONING_OUT_CMDLEN,
                                   false, sizeof(d), d));
    }

    ptvp = construct_scsi_pt_obj();
    if (NULL == ptvp) {
        pr2serr("%s: out of memory\n", b);
        return -1;
    }
    set_scsi_pt_cdb(ptvp, zo_cdb, sizeof(zo_cdb));
    set_scsi_pt_sense(ptvp, sense_b, sizeof(sense_b));
    res = do_scsi_pt(ptvp, sg_fd, tmo, verbose);
    ret = sg_cmds_process_resp(ptvp, b, res, noisy,
                               verbose, &sense_cat);
    if (-1 == ret) {
        if (get_scsi_pt_transport_err(ptvp))
            ret = SG_LIB_TRANSPORT_ERROR;
        else
            ret = sg_convert_errno(get_scsi_pt_os_err(ptvp));
    } else if (-2 == ret) {
        switch (sense_cat) {
        case SG_LIB_CAT_RECOVERED:
        case SG_LIB_CAT_NO_SENSE:
            ret = 0;
            break;
        default:
            ret = sense_cat;
            break;
        }
    } else
        ret = 0;
    destruct_scsi_pt_obj(ptvp);
    return ret;
}


int
main(int argc, char * argv[])
{
    bool all = false;
    bool close = false;
    bool finish = false;
    bool open = false;
    bool quick = false;
    bool reamz = false;
    bool element_id_given = false;
    bool sequentialize = false;
    bool verbose_given = false;
    bool version_given = false;
    int res, c, n;
    int sg_fd = -1;
    int tmo = DEF_PT_TIMEOUT;
    int verbose = 0;
    int ret = 0;
    int sa = 0;
    uint16_t zc = 0;
    uint64_t zid = 0;
    int64_t ll;
    const char * device_name = NULL;
    const char * sa_name;

    while (1) {
        int option_index = 0;

        c = getopt_long(argc, argv, "acC:e:fhoqrRSt:vVz:", long_options,
                        &option_index);
        if (c == -1)
            break;

        switch (c) {
        case 'a':
        case 'R':
            all = true;
            break;
        case 'c':
            close = true;
            sa = CLOSE_ZONE_SA;
            break;
        case 'C':
            n = sg_get_num(optarg);
            if ((n < 0) || (n > 0xffff)) {
                pr2serr("--count= expects an argument between 0 and 0xffff "
                        "inclusive\n");
                return SG_LIB_SYNTAX_ERROR;
            }
            zc = (uint16_t)n;
            break;
        case 'e':
            ll = sg_get_llnum(optarg);
            if ((ll < 0) || (ll > UINT32_MAX)) {
                pr2serr("bad argument to '--element=EID'\n");
                return SG_LIB_SYNTAX_ERROR;
            }
            if (0 == ll)
                pr2serr("Warning: 0 is an invalid element identifier\n");
            zid = (uint64_t)ll;    /* putting element_id in zid */
            element_id_given = true;
            break;
        case 'f':
            finish = true;
            sa = FINISH_ZONE_SA;
            break;
        case 'h':
        case '?':
            usage();
            return 0;
        case 'o':
            open = true;
            sa = OPEN_ZONE_SA;
            break;
        case 'q':
            quick = true;
            break;
        case 'r':
            reamz = true;
            sa = REM_ELEM_MOD_ZONES_SA;
            break;
        case 'S':
            sequentialize = true;
            sa = SEQUENTIALIZE_ZONE_SA;
            break;
        case 't':
            tmo = sg_get_num(optarg);
            if (tmo < 0) {
                pr2serr("bad argument to '--timeout='\n");
                return SG_LIB_SYNTAX_ERROR;
            }
            break;
        case 'v':
            verbose_given = true;
            ++verbose;
            break;
        case 'V':
            version_given = true;
            break;
        case 'z':
            ll = sg_get_llnum(optarg);
            if (-1 == ll) {
                pr2serr("bad argument to '--zone=ID'\n");
                return SG_LIB_SYNTAX_ERROR;
            }
            zid = (uint64_t)ll;
            break;
        default:
            pr2serr("unrecognised option code 0x%x ??\n", c);
            usage();
            return SG_LIB_SYNTAX_ERROR;
        }
    }
    if (optind < argc) {
        if (NULL == device_name) {
            device_name = argv[optind];
            ++optind;
        }
        if (optind < argc) {
            for (; optind < argc; ++optind)
                pr2serr("Unexpected extra argument: %s\n",
                        argv[optind]);
            usage();
            return SG_LIB_SYNTAX_ERROR;
        }
    }

#ifdef DEBUG
    pr2serr("In DEBUG mode, ");
    if (verbose_given && version_given) {
        pr2serr("but override: '-vV' given, zero verbose and continue\n");
        /* verbose_given = false; */
        version_given = false;
        verbose = 0;
    } else if (! verbose_given) {
        pr2serr("set '-vv'\n");
        verbose = 2;
    } else
        pr2serr("keep verbose=%d\n", verbose);
#else
    if (verbose_given && version_given)
        pr2serr("Not in DEBUG mode, so '-vV' has no special action\n");
#endif
    if (version_given) {
        pr2serr("version: %s\n", version_str);
        return 0;
    }

    if (1 != ((int)close + (int)finish + (int)open + (int)sequentialize +
              (int)reamz)) {
        pr2serr("One, and only one, of these options needs to be given:\n"
                "   --close, --finish, --open, --remove or --sequentialize "
                "\n\n");
        usage();
        return SG_LIB_CONTRADICT;
    }
    if (element_id_given && (! reamz)) {
        pr2serr("The --element=EID option should only be used with the "
                "--remove option\n\n");
        usage();
        return SG_LIB_CONTRADICT;
    }
    sa_name = sa_name_arr[sa];

    if (0 == tmo)
        tmo = DEF_PT_TIMEOUT;
    if (NULL == device_name) {
        pr2serr("missing device name!\n");
        usage();
        return SG_LIB_SYNTAX_ERROR;
    }

    sg_fd = sg_cmds_open_device(device_name, false /* rw */, verbose);
    if (sg_fd < 0) {
        int err = -sg_fd;
        if (verbose)
            pr2serr("open error: %s: %s\n", device_name,
                    safe_strerror(err));
        ret = sg_convert_errno(err);
        goto fini;
    }
    if (reamz && (! quick)) {
        if (! element_id_given)
            pr2serr(">>> Warning: --element=EID option NOT given so "
                    "assuming element 0\n");
        sg_warn_and_wait(sa_name_arr[REM_ELEM_MOD_ZONES_SA], device_name,
                         false);
    }

    res = sg_ll_zone_out(sg_fd, sa, zid, zc, all, tmo, true, verbose);
    ret = res;
    if (res) {
        if (SG_LIB_CAT_INVALID_OP == res)
            pr2serr("%s command not supported\n", sa_name);
        else {
            char b[80];

            sg_get_category_sense_str(res, sizeof(b), b, verbose);
            pr2serr("%s command: %s\n", sa_name, b);
        }
    }

fini:
    if (sg_fd >= 0) {
        res = sg_cmds_close_device(sg_fd);
        if (res < 0) {
            pr2serr("close error: %s\n", safe_strerror(-res));
            if (0 == ret)
                ret = sg_convert_errno(-res);
        }
    }
    if (0 == verbose) {
        if (! sg_if_can2stderr("sg_zone failed: ", ret))
            pr2serr("Some error occurred, try again with '-v' or '-vv' for "
                    "more information\n");
    }
    return (ret >= 0) ? ret : SG_LIB_CAT_OTHER;
}
